support optional tag for pointer fields

yunwu-jnpr opened the pull request:
Jira: MIST-101148

yunwu-jnpr committed:
support optional tag for pointer fields

nsd20463 reviewed (approved):
Excellent unit testing.
I'm going to make one change to how p.asProtobuf is created to make it less repetitive and merge it.
Thanks.

yunwu-jnpr commented:
Thanks, @nsd20463, appreciate your help, always

nsd20463 commented:
What should happen if ",optional" is added to a non-pointer field in the Go struct definition?
Is it an error?
Is "optional" copied through to the .proto output?
Either way is valid.

yunwu-jnpr commented:
> What should happen if ",optional" is added to a non-pointer field in the Go struct definition? Is it an error? Is "optional" copied through to the .proto output?

For non-pointer field, `optional` tag is ignored.
Hmm... Maybe we should print an error, so user can take action -- either remove `optional` for that field, or change the field to pointer field.

nsd20463 commented:
So you think they should make the field a pointer? The way I think about it, it doesn't have to be a pointer here, just in the protoc generated stub. "optional" flag is not (and cannot) be handled by this protobuf3.Unmarshal(). "optional" is a flag to be passed through the .proto file to the `protoc` compiler, so that it generates stubs which support optional behavior.

Also, `map` type is also a pointer internally, and you can tell the difference between nil and empty-but-not-nil maps.
(and slices too for that matter---and they too have a pointer inside them)

So I'm changing the code so "optional" can be applied anywhere, and adding a comment saying that, of course, it cannot apply to the Go struct, but it will apply to stubs generated from the .proto.

yunwu-jnpr commented:
Sure, of course. Thanks for making it better!

nsd20463 committed:
gofmt

nsd20463 committed:
allow ",optional" to apply to all data types

The optional flag is something we pass through to protoc by copying it
into the protobuf definitions we generate. So it can apply to any type,
not just pointer types.

Also a map is, internally, a pointer type, and you can distinguish
between empty and nil maps. So even without the above choice, map
(and slice) would still need to be handled.

Once "optional" applies to all types, I can simplify how the protobuf
is generated. And finally I added to the unit tests to check a
non-pointer and a map.
